### ISSUE TITLE
util/variant: Add an indexed reader

### ISSUE THREAD

**Issue by zaeleus** [zaeleus/noodles#188]
> Is there an abstraction / trait that unifies reading from an indexed BCF or VCF where the index is either CSI or TBI?

_Originally posted by @brentp in https://github.com/zaeleus/noodles/discussions/170#discussioncomment-6529163_

---

E.g.,

```rust
let mut reader = variant::indexed_reader::Builder::default().build_from_path("src.vcf.gz")?;
let header = reader.read_header()?;
let region = Region::new("sq0", ..);

for result in reader.query(&header, &region)? {
    let record = result?;
    // ...
}
```

**Comment by zaeleus** [comment]
[`variant::IndexedReader`](https://docs.rs/noodles-util/0.18.0/noodles_util/variant/indexed_reader/enum.IndexedReader.html) is now available in noodles-util 0.18.0. See [`noodles-util/examples/util_variant_query.rs`](https://github.com/zaeleus/noodles/blob/noodles-0.46.0/noodles-util/examples/util_variant_query.rs) for an example.

**Comment by brentp** [comment]
Thanks very much!

**Comment by zaeleus** [comment]
Not yet implemented due to the BCF's indexed reader builder not having it implemented either. I'll work on adding this functionality this week.

**Comment by zaeleus** [comment]
The detection was indeed discarding the start of file buffer, sorry! This is fixed (with test) in noodles-util 0.19.1. Thank you for both your feedback and patience .
